Use `UTxO era` as single source of truth in balanceTransaction

Anviking opened the pull request:
- [x] Add `eraHistory` getter to `NetworkLayer`. Unfortunately we cannot create it from a `TimeInterpreter`. 
- [x] More consistently use `Cardano.UTxO` for input resolution in balanceTx
    1. Make `evaluateTransactionBalance` take `Cardano.UTxO` directly
    2. Make `assignScriptRedeemers` take `Cardano.UTxO`, instead of manually converting to `Alonzo.UTxO`
    3. Make `balanceTx` take `Cardano.UTxO` as part of `PartialTx` field `inputs`.
    4. Let the `Cardano.UTxO` be created in the server handler directly from Api types
- [x] Improve doc comments for balanceTransaction
- [ ] Use `Ledger.UTxO` instead!

### Comments

#### Q: Do we really need a new `eraHistory` function in the NetworkLayer?

Alternatively we could add the following constraint to `TimeInterpreter`

```patch
-data TimeInterpreter m = forall eras. TimeInterpreter
-    { interpreter :: m (Interpreter eras)
+data TimeInterpreter m = forall xs. Cardano.ConsensusBlockForMode Cardano.CardanoMode ~ Consensus.HardForkBlock xs => TimeInterpreter
+    { interpreter :: m (Interpreter xs)
```

with the downside of not being able to as easily call `mkTimeInterpreter` with e.g. two eras like in https://github.com/input-output-hk/cardano-wallet/blob/50a100e74f404a948a18cbecc236eb924cdf9ee2/lib/core/test/unit/Cardano/Wallet/Primitive/SlottingSpec.hs#L221-L237, we'd need to use empty eras as a hack, like [here](https://github.com/input-output-hk/cardano-wallet/blob/50a100e74f404a948a18cbecc236eb924cdf9ee2/lib/core/src/Cardano/Wallet/Primitive/Slotting.hs#L670-L685).

### Issue Number

ADP-1662

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -611,3 +634,54 @@ unsafeExtendSafeZone = f . neverFails r
+-------------------------------------------------------------------------------}
+
+dummyEraHistory :: (Cardano.EraHistory Cardano.CardanoMode, SystemStart)
+dummyEraHistory =
```
No longer being able to test balanceTx with `mkSingleEraInterpreter` and having to add a separate `eraHistory` to `NetworkLayer` — I'm not happy about it. It does work, however, and might be preferable to the alternatives...

Anviking committed:
Add `eraHistory` getter to `NetworkLayer`

In order to switch from `Alonzo.evaluateTransactionExecutionUnits` to
`Cardano.evaluateTransactionExecutionUnits` we need an `EraHistory`.
Because `TimeInterpreter` can internally contain an EraSummary of any
shape (forall xs), and `EraHistory` only works with a closed set of
modes, we cannot convert between the two, and thus need a new accessor
to the `NetworkLayer`.

There should in principle exist several ways to reduce this duplication
in the future.

Includes squashed review suggestions.

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Anviking committed:
Add dummyEraHistory

With squashed review feedback from Jonathan.

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Anviking committed:
Tweak Address generator to create valid addresses

Anviking committed:
Improve doc comments for balanceTransaction

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network/Node.hs:
```diff
@@ -374,6 +376,10 @@ withNodeNetworkLayerBase tr net np conn versionData tol action = do
             _timeInterpreter (contramap MsgInterpreterLog tr) interpreterVar
+        , eraHistory = do
+            i <- atomically (readTMVar interpreterVar)
+            let W.StartTime t0 = getGenesisBlockDate
```
We _could_ replace our own `StartTime` with `Cardano.Slotting.Time.SystemStart` from `cardano-base` to avoid conversions like these. We are getting our `SlotNo` from there. (Though not too keen to do so now.)

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1545,14 +1544,40 @@ balanceTransaction
+    -- present in the transaction, ensuring collateral is never forfeited.
+    --
+    -- TODO: Remove 'W.ProtocolParameters' argument.
+    -> (Cardano.EraHistory Cardano.CardanoMode, SystemStart)
```
The `SystemStart` _is_ duplicated with `getGenesisBlockDate` in `W.ProtocolParameters` which we'd like to replace, but _not_ `Cardano.ProtocolParameters`.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1267,7 +1256,7 @@ _assignScriptRedeemers (toAlonzoPParams -> pparams) ti resolveInput redeemers tx
                 { Alonzo.body = (Alonzo.body alonzoTx)
                     { Alonzo.scriptIntegrityHash = Alonzo.hashScriptIntegrity
-                        pparams
+                        (toAlonzoPParams pparams)
```
Will need some adaptation when we enable Vasil, but that's a problem for another PR.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs:
```diff
@@ -163,6 +165,8 @@ import UnliftIO.Async
     ( Exception )
 
+import qualified Cardano.Api as Cardano
+
```
https://input-output-hk.github.io/adrestia/code/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

```suggestion
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs:
```diff
@@ -255,6 +260,12 @@ withNetworkLayer tr network np project k =
             pure $ HF.mkInterpreter $ networkSummary networkId
 
+    eraHistoryFromStartTime ::
+        StartTime -> IO (Cardano.EraHistory Cardano.CardanoMode, SystemStart)
```
Unrelated to this PR, but this module deviates from our coding standard, which indicates the following style when wrapping type signatures:

```haskell
mkMagicalBlock
    :: MagicProtocolId
    -> MagicType
    -> MagicalKey
    -> Maybe Delegation.MagicalCertificate
    -> Block
```

I suggest that we use the standard style for this new function:

```suggestion
    eraHistoryFromStartTime
        :: StartTime
        -> IO (Cardano.EraHistory Cardano.CardanoMode, SystemStart)
```

(Incidentally, the indentation in our [coding standards](https://input-output-hk.github.io/adrestia/code/Coding-Standards#strategy-2-place-comments-on-their-own-line-instead-of-attempting-to-align-them-vertically) document seems to have been corrupted. I'll see if I can fix it!)

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -611,3 +632,55 @@ unsafeExtendSafeZone = f . neverFails r
         }
     r = "unsafeExtendSafeZone should make PastHorizonExceptions impossible."
+
+
```
https://input-output-hk.github.io/adrestia/code/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

```suggestion

```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -611,3 +632,55 @@ unsafeExtendSafeZone = f . neverFails r
+            { eraEpochSize = EpochSize 432000
+            , eraSlotLength = mkSlotLength 1
+            , eraSafeZone = StandardSafeZone 129600
+            }
```
Use 4-space indents:

```suggestion
    eraParams = EraParams
        { eraEpochSize = EpochSize 432000
        , eraSlotLength = mkSlotLength 1
        , eraSafeZone = StandardSafeZone 129600
        }
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -611,3 +632,55 @@ unsafeExtendSafeZone = f . neverFails r
+                , boundEpoch = Cardano.EpochNo 0
+                }
+            , eraParams = eraParams
+            }
```
4-space indents:

```suggestion
    emptyEra = EraSummary
        { eraStart = Bound
            { boundTime = RelativeTime 0
            , boundSlot = 0
            , boundEpoch = Cardano.EpochNo 0
            }
        , eraEnd = EraEnd Bound
            { boundTime = RelativeTime 0
            , boundSlot = 0
            , boundEpoch = Cardano.EpochNo 0
            }
        , eraParams = eraParams
        }
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -611,3 +632,55 @@ unsafeExtendSafeZone = f . neverFails r
+                    , eraEnd = EraUnbounded
+                    , eraParams = eraParams
+                    }
+            }
```
```suggestion
    summary = Summary
        { getSummary =
            NonEmptyCons emptyEra -- Byron
            $ NonEmptyCons emptyEra -- Shelley
            $ NonEmptyCons emptyEra -- Allegra
            $ NonEmptyCons emptyEra -- Mary
            $ NonEmptyOne EraSummary -- Alonzo
                { eraStart = Bound
                    { boundTime = RelativeTime 0
                    , boundSlot = 0
                    , boundEpoch = Cardano.EpochNo 0
                    }
                , eraEnd = EraUnbounded
                , eraParams = eraParams
                }
        }
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/Address/Gen.hs:
```diff
@@ -88,4 +88,4 @@ data Parity = Even | Odd
 
 mkAddress :: Char -> Address
-mkAddress c = Address $ "ADDR" `B8.snoc` c
+mkAddress c = Address $ BS.pack (97 : replicate (28 - 1) 0) `B8.snoc` c
```
Hi @Anviking 

Would it be possible to replace these magic constants with **_named constants_**, perhaps in a `where` clause?

We could also give each named constant a comment stating its provenance (where we got it from).

This way a reader can understand why we've chosen to use 97 and 28 here, rather than some other values.

https://input-output-hk.github.io/adrestia/code/Coding-Standards#prefer-named-constants-over-magic-numbers

Anviking replied:
0cd0bddf6ec9d59aed3ee5a5e4f69882e6f0c3d3

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -652,6 +644,10 @@ newTransactionLayer networkId = TransactionLayer
+
+    , toCardanoUTxO = _toCardanoUTxO
+    , _fromCardanoTxOut = fromCardanoTxOut
+    , _fromCardanoTxIn = fromCardanoTxIn
```
Suggestion: remove these underscores, for consistency with other functions within this record.

See: 2e2126978f4ba0550af5a1d905ae79cd7b1b1d2b.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2770,21 +2761,6 @@ instance Arbitrary Wallet' where
-            simplifyUTxO = UTxO . uncurry Map.singleton
-                . foldl1 (\(i, TxOut addr _val1) (_, TxOut _ _val2)
-                    -> (i, TxOut addr (TokenBundle.fromCoin (Coin 10_000_000))))
-                . UTxO.toList
```
👍🏻

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2795,27 +2771,47 @@ instance Buildable a => Show (ShowBuildable a) where
+        | addr' <- prependOriginal shrink addr
+        , val' <- prependOriginal shrink val
+        , dat' <- prependOriginal shrink dat
+        ]
```
@Anviking 

Just curious, have we tested this shrinker anywhere?

Alternatively, have you considered using `shrinkMapBy`?
**See 818c4fb64d9b3bb82ef02baf0155a9cb50b65adf.**

```suggestion
    shrink = shrinkMapBy
        (Compatibility.toCardanoTxOut ShelleyBasedEraAlonzo)
        (Compatibility.fromCardanoTxOut)
        (shrink)
```

Of course, this relies on the following property, which may or may not be true (I haven't checked):

```hs
toCardanoTxOut era . fromCardanoTxOut == id
```

Anviking replied:
> Of course, this relies on the following property, which may or may not be true (I haven't checked):

Unfortunately not 😄🥲

The wallet types don't know about datum hashes, and won't know about Vasil inline datums or scripts.

Anviking replied:
> Just curious, have we tested this shrinker anywhere?

I tried it yesterday, and it seemed to work in that case.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2795,27 +2771,47 @@ instance Buildable a => Show (ShowBuildable a) where
         let (Cardano.Tx (Cardano.TxBody content) _) = tx
         let inputs = Cardano.txIns content
-        resolvedInputs <- forM inputs $ \i -> do
+        resolvedInputs <- Cardano.UTxO . Map.fromList <$> forM inputs (\i -> do
```
**Suggestion**: Use `BlockArguments`, as it would allow us to remove the parentheses from around the monadic expression.

**See bcd3b4541112020f98700af332247f470ca01b10.**

```suggestion
        resolvedInputs <- Cardano.UTxO . Map.fromList <$> forM inputs \i -> do
```

We would then not need to "smuggle in" the trailing bracket a few lines later.

HeinrichApfelmus replied:
Unrelated: @jonathanknowles , could you record this example as an argument in favor of `BlockArguments`? Personally, I'm not yet convinced by this extension (does `Cardano.UTxO . Map.fromList <$> forM inputs \i -> i` parse?) — regardless, I think that the use or unuse of `BlockArguments` should be discussed with the team so that we have a common Haskell idiom that we share.

(One reason for me insisting on a shared Haskell idiom is past use of lenses in the wallet — if I understood correctly, that didn't work out so well, and I believe that's because they were not explicitly made part of a shared idiom. (Even more unrelated — happy to discuss outside of this PR).)

Anviking replied:
🤔 Ah, thanks! But I will not apply this before there is team consensus then.

> Does `Cardano.UTxO . Map.fromList <$> forM inputs \i -> i parse?`

No:

<img width="831" alt="Skärmavbild 2022-05-13 kl  16 21 12" src="https://user-images.githubusercontent.com/304423/168304142-68f527c1-2348-4006-a9d2-b27b37dde7fd.png">

Often you can avoid parenthesis by flipping the order of things, but I didn't find that to be possible here (and a few similar cases I have run into)

jonathanknowles replied:
> (does Cardano.UTxO . Map.fromList <$> forM inputs \i -> i parse?

@HeinrichApfelmus here is a commit that compiles: https://github.com/input-output-hk/cardano-wallet/commit/bcd3b4541112020f98700af332247f470ca01b10

> I think that the use or unuse of BlockArguments should be discussed with the team so that we have a common Haskell idiom that we share.

Fair enough! I started a poll in our development channel [here](https://element.adrestia.iohkdev.io/#/room/!pVmOFJiBkItJMNTmCR:adrestia.iohkdev.io/$SUQD44U_30OQl_hpz2dZUGyBoRnz_8M9Ng6-FCN5cTQ?via=adrestia.iohkdev.io).

I personally don't have a strong feeling either way about `BlockArguments`, but I do think that `do` blocks can become a bit unwieldy (and difficult to follow) when they are are enclosed within parentheses spanning multiple lines. In this case, the trailing brace is rather difficult to spot in the GitHub viewer. (Though with a real editor, we'd have matching brace highlighting to help us.)

If we couldn't use `BlockArguments`, then I'd advocate for extracting out the "block" to make it a separate value in cases like these.

HeinrichApfelmus replied:
> but I do think that do blocks can become a bit unwieldy (and difficult to follow) when they are are enclosed within parentheses spanning multiple lines.

The traditional solution to that is to have the `do` block be the last argument of the function, and slap a `$` before it, e.g.

```
foo arg1 arg2 $ \i -> do
    something
```

However, this solution does not work in this case, because of the fixity of the operator `<$>` is higher than the fixity of `$`, so `<$>` binds tigher. Put differently, the expression

```
Map.fromList <$> forM input $ \i -> do {…}
```

is parsed as

```
(Map.fromList <$> forM input) $ \i -> do {…}
```

but we want this:

```
Map.fromList <$> (forM input $ \i -> do {…})
```

My personal go-to solution is to use `fmap` instead of the operator:

```
fmap Map.fromList . forM input $ \i -> do {…}

```

> Fair enough! I started a poll in our development channel here

Thanks! My proposal would be to discuss and move this forward this during our next dev meeting as well. (It was on the agenda already, sorry for postponing.)

Anviking replied:
> My personal go-to solution is to use fmap instead of the operator:

🤔 I thought I tried that, but apparently not! 👍

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2824,22 +2820,8 @@ instance Arbitrary (PartialTx Cardano.AlonzoEra) where
-    convertUTxO (_, _, Just _) =
-        error "resolvedInputsUTxO: todo: handle datum hash"
+        -- TODO: Add a shrinker
+        shrinkUTxOPreserveShape (Cardano.UTxO _) = []
```
@Anviking, do you plan to add a shrinker in this PR?

(I guess it's difficult to claim that it preserves the shape if it replaces every UTxO set with the empty set! 😄)

Anviking replied:
> if it replaces every UTxO set with the empty set! 😄)

Oh but that it doesn't do! The claim would be vacuously true when it doesn't actually shrink 😄 

> do you plan to add a shrinker in this PR?

🤔 I probably should have another look here, _however_ in general I am happy with taking on some short-to-medium-term technical debt in this PR regarding the shrinkers (and to some extent property test coverage, and generators), in the progress of moving from wallet types to `Cardano` ones.

Anviking replied:
Some improvements: https://github.com/input-output-hk/cardano-wallet/pull/3255/commits/d524718044a91759ccaac54f9e75e9dc97398bb7

jonathanknowles reviewed (commented):
Hi @Anviking 

Thanks for making this PR.

I've reviewed some of it (see comments and [branch with suggestions](https://github.com/input-output-hk/cardano-wallet/commits/jonathanknowles/ADP-1662/EraHistory)), but I still have a few more things to look at.

I'll post another review shortly!

All the best

Jonathan

Anviking committed:
Remove outdated TODO

It was written in October 2021. We have since moved most of the
implementation to Wallet.hs. The remaining logic in Server.hs
is likely -- mostly -- here to stay.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1545,14 +1548,43 @@ balanceTransaction
+    -- a local node.
+    --
+    -- Providing incorrect values will result in a phase 1 script integrity
+    -- hash failures instead of phase 2 failures if script redeemers are
```
```suggestion
    -- Providing incorrect values will result in phase 1 script integrity
    -- hash failures instead of phase 2 failures if script redeemers are
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1759,27 +1794,61 @@ balanceTransactionWithSelectionStrategy
+        unless (null conflicts) $
+            throwE ErrBalanceTxConflictingInputResolution
+      where
+         unUTxO (Cardano.UTxO u) = u
```
Hi @Anviking!

I find this function a little difficult to understand, especially at first glance.

Though more importantly, I suspect it's difficult to test that the combining behaviour is correct in isolation, because we have to call it via `balanceTransaction`.

Given that we're guarding the correctness of `combinedUTxO`, perhaps we could extract out both the definition of `combinedUTxO` and `guardWalletUTxOConsistencyWith` into a separate **_pure_** function that is easy to test?

The function would take a pair of `Cardano.UTxO` values, and return a combined `Cardano.UTxO` if and only if there are no inconsistencies.

I've written a prototype here: 8866ffde6222944a69ba18a278816409e49920fa.

Since we don't care about reporting which values are in conflict (at the moment), we can use simple set operations to determine if there are conflicts.

Anviking replied:
I hear you, and it did cross my mind, but I'm not sure.

My main reservation is: it's incompatible with a global `combinedUTxO` `where` binding. It's crucial for helpers like `balanceAfterSettingMinFee`, `guardTxBalance`, `assembleTransaction` to use the same `Cardano.UTxO`. With a `combinedUTxO <- combineWalletUTxOAndExternalUTxO` we'd need to either manually pass the `utxo` around, which adds noise and room for mistakes, or add more indirection like increased nesting or an internal "Balance monad". The latter I'm curious about, but now's the wrong time to think about it.

Extended prototype showing how it would turn out -> 2eaab252d1a192b63a4ffdb8e92ccd4ecedc4d01.

> Though more importantly, I suspect it's difficult to test that the combining behaviour is correct in isolation, because we have to call it via balanceTransaction.

1. `prop_balanceTransactionValid` currently fails without `guardWalletUTxOConsistencyWith`. The guard was added because of a dropped [hack](https://github.com/input-output-hk/cardano-wallet/pull/3255/files#diff-45f59e6a8905a1e2f6cc28025e3c5e5ab6e37b9c79735f222569e677d5d99ca2L3362-L3375) which artificially made the test data consistent. (Sorry, this was not easily apparent from the commit history)
2. If possible, I would much prefer to focus future tests for this on the `balanceTransaction` interface. (E.g. if the script inputs I'm trying to redeem exist in my wallet utxo, I should still be able to redeem them.)

So thanks for raising this and for the prototype (I do want to move to Cardano.UTxO types like you did there), but could we ignore this for now?

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -4521,7 +4521,11 @@ instance IsServerError ErrBalanceTx where
+            apiError err403 ConflictingInputResolution $ mconcat
+                [ "The provided input resolution conflicts with my UTxO. Please"
+                , " make sure it is correct."
+                ]
```
```suggestion
                [ "At least one of the inputs provided has a different value "
                , "or different address from that recorded in the wallet UTxO. "
                , "Please ensure it is correct."
                ]
```

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -4768,6 +4768,18 @@ x-errExistingKeyWitnesses: &errExistingKeyWitnesses
+    message:
+      type: string
+      description: |
+        Occurs when providing an input resolution which conflicts with the wallet UTxO set when balancing.
```
I think the name `conflicting_input_resolution` is a bit confusing, because I'm not sure what a `resolution` is.

Perhaps just `conflicting_input`?

Anviking replied:
The thing is that it's the provided (input -> output) map (essentially an UTxO set) which conflicts with the wallet UTxO set. I think `conflicting_input_resolution` is more clear than  `conflicting_input `. However, I can tweak the wording a bit... -> 0aa2222b4e794c5151409e56949efcbd8d46996c

jonathanknowles replied:
Can I suggest some alternatives:
- `conflicting_input`
- `conflicting_input_output_mapping`
- `conflicting_input_output_mappings`
- `conflicting_utxo`
- `conflicting_utxos`

Question: do we, on detection of an erroneous entity (of which there may be many), usually return a singular error, or a plurality?

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -4768,6 +4768,19 @@ x-errExistingKeyWitnesses: &errExistingKeyWitnesses
       type: string
       enum: ['existing_key_witnesses']
 
+x-errConflictingInputResolution: &errConflictingInputResolution
+  <<: *responsesErr
+  title: conflicting_input_resolution
+  properties:
+    message:
+      type: string
+      description: |
+        Occurs when providing an input resolution (input -> output mapping) as part of the
+        `inputs` field which conflicts with the wallet UTxO set when balancing.
+    code:
+      type: string
+      enum: ['conflicting_input_resolution']
```
Hi @Anviking 

In the API specification, I think we should try hard to avoid using phrases like:

- "providing an input resolution"
- "a conflicting input resolution"
- "there were two conflicting input resolutions"

It could just be me, but I feel the above usages of "resolution" sound very awkward in English.

The following phrases sound much more natural to me:

- "a conflicting input"
- "an external input that is inconsistent with an input in the wallet's UTxO set"
- "there were two external inputs that are inconsistent with inputs recorded in the wallet's UTxO set"

I would much prefer to see a wording like this:

```suggestion
x-errConflictingInput: &errConflictingInput
  <<: *responsesErr
  title: conflicting_input
  properties:
    message:
      type: string
      description: |
        Occurs when the additional transaction `inputs` field contains an
        input that is also present in the wallet's UTxO set, but where the
        address or value recorded in the wallet's UTxO set differs from
        the address or value provided.
    code:
      type: string
      enum: ['conflicting_input']
```

With this wording, we can explain exactly what causes the conflict, without having to use a phrasing that sounds awkward in English.

If the "conflicting input resolution" terminology was just used within the code and never exposed to the user, then I think it wouldn't matter so much.

But because this adds a new named type to the API (which is harder to change after we've made a release), I think it's worth taking the time to gain consensus with this.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3275,6 +3297,9 @@ prop_balanceTransactionValid wallet (ShowBuildable partialTx') seed
             Left ErrBalanceTxMaxSizeLimitExceeded ->
                 label "maxTxSize limit exceeded" $ property True
+            Left ErrBalanceTxConflictingInputResolution ->
+                 label "input resolution conflicts with wallet utxo"
```
```suggestion
                label "input resolution conflicts with wallet utxo"
```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3275,6 +3297,9 @@ prop_balanceTransactionValid wallet (ShowBuildable partialTx') seed
                 label "maxTxSize limit exceeded" $ property True
+            Left ErrBalanceTxConflictingInputResolution ->
+                 label "input resolution conflicts with wallet utxo"
+                    $ property True
```
Hi @Anviking 

I think it's great that we're categorizing the different failure cases here. 👍🏻 

However, I'm concerned that we don't currently appear to be checking for completeness of coverage.

If we use `label` inside a _pattern match_ of a _sum type_, I think we have no guarantee that each branch of the pattern will actually match during a test suite run. _If_ a particular branch matches, then the label will be added to the test suite output. But if a branch _doesn't_ match, then the only way to notice would be to _manually_ inspect the output of a test suite run and notice that the corresponding label is missing.

Ideally, I think we want some kind of _automated_ warning about coverage failures. Given that we're using QuickCheck, I think we'd want to:
- use some combination of [`checkCoverage`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#v:checkCoverage) and [`cover`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#v:cover) or [`coverTable`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#v:coverTable);
- ensure that the coverage check is made _outside_ of the pattern match.

There doesn't seem to be an ideal way to do this, but we could use a method similar to this:

1. For the cases we care about, define a series of functions from `ResultType` to `Bool`: 
https://github.com/input-output-hk/cardano-wallet/blob/887ce41c3bf1c049257fa1e09cbf134822c8e415/lib/core/test/unit/Cardano/Wallet/CoinSelection/InternalSpec.hs#L268-L273

2. Place [`cover`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#v:cover) statements outside of the pattern match:
https://github.com/input-output-hk/cardano-wallet/blob/887ce41c3bf1c049257fa1e09cbf134822c8e415/lib/core/test/unit/Cardano/Wallet/CoinSelection/InternalSpec.hs#L241-L246

3. Define an auxiliary function to provide us with an exhaustiveness check, in case someone adds another case to the sum type:
https://github.com/input-output-hk/cardano-wallet/blob/887ce41c3bf1c049257fa1e09cbf134822c8e415/lib/core/test/unit/Cardano/Wallet/CoinSelection/InternalSpec.hs#L300

This method does add a bit of boilerplate, but at least it provides us with assurance that we've covered the cases we care about.

There might be a better method available with some clever use of [`coverTable`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#v:coverTable).

Regardless, I think we ought to use _some_ method of checking coverage here, otherwise we leave ourselves open the possibility of silent regressions in either the code or the test suite.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/Address/Gen.hs:
```diff
@@ -88,4 +89,12 @@ data Parity = Even | Odd
+    BS.pack [mainnetEnterpriseAddrTag] <> B8.pack dummyPaymentKeyHash
+    where
+      mainnetEnterpriseAddrTag = 0b01100001
+      -- https://cips.cardano.org/cips/cip19
```
Thank-you for naming these constants! 👍🏻

jonathanknowles reviewed (commented):
Hi @Anviking 

Thanks again for making this PR.

There is something that I'd like us to address relating to the API specification. In particular, I think we can simplify the terminology we use when reporting conflicting inputs (see [comment](https://github.com/input-output-hk/cardano-wallet/pull/3255#discussion_r886351315)). I think it's important to get this right before we merge, because it's harder to change the API once we've already included this change in a release.

I've also made a suggestion about adding coverage checks to the `balanceTransaction` property tests (see [comment](https://github.com/input-output-hk/cardano-wallet/pull/3255#discussion_r886358943)), to make sure that we cover the various error conditions that can occur. If we don't add coverage checks in this PR, then I think we should record this as an item of engineering debt and make a ticket for it. (EDIT: I've created ADP-1843 to record this.) However, given that the property concerned doesn't have any coverage checks in `master`, I think this PR still represents a strict improvement over what we currently have, and so it needn't be a block against merging.

I'm more concerned with the API specification changes. Perhaps we can discuss in the channel?

Thanks again for making this PR, and for your patience in waiting for a review!

Jonathan

Anviking committed:
Update balanceTx goldens to Babbage txs

Anviking committed:
Move test data for pingPong to top level

Anviking committed:
Improve testing of error cases in balanceTx

- Unresolved inputs
- Validity interval past horizon
- Faulty redeemers

Anviking committed:
Split off distributeSurplusSpec

Anviking committed:
Add dummyForkEraHistory

Anviking committed:
Add more conversion functions

Anviking committed:
Rely more on cardano-api in balanceTx

Core goal: replace ad-hoc output/utxo type with `Cardano.UTxO` here

```
 data PartialTx era = PartialTx
     { tx :: Cardano.Tx era
-    , inputs :: [(TxIn, TxOut, Maybe (Hash "Datum"))]
+    , inputs :: Cardano.UTxO era
```

How?

- Convert from API types to `Cardano.UTxO` once in server handler
- Bonus: add `guardWalletUTxOConsistencyWith inputUTxO`
  - Need is verified by property test.
- Make `assignScriptRedeemers` use cardano-api functions rather than
  ledger functions internaly.
  - This requires changes to our error types

Anviking committed:
fixup warnings

Anviking committed:
Update documentation for PartialTx
